File: apps/web/src/api/listMeta.ts
/** DEV13-D：解析控制面列表分页响应头。本页条数不得冒充全量。 */

export const HDR_LIMIT = 'x-siq-list-limit';
export const HDR_RETURNED = 'x-siq-list-returned';
export const HDR_TRUNCATED = 'x-siq-list-truncated';
export const HDR_NEXT_CURSOR = 'x-siq-next-cursor';
export const HDR_TOTAL = 'x-siq-list-total';

export interface ListMeta {
  limit: number | null;
  returned: number | null;
  /** true=还有后续页；缺失头时视为未知（不得宣称「共 returned 条」） */
  truncated: boolean | null;
  nextCursor: string | null;
  total: number | null;
}

function parseIntHeader(headers: Headers, name: string): number | null {
  const raw = headers.get(name);
  if (raw == null || raw === '') return null;
  const n = Number.parseInt(raw, 10);
  return Number.isFinite(n) ? n : null;
}

export function parseListMeta(headers: Headers): ListMeta {
  const truncRaw = headers.get(HDR_TRUNCATED);
  let truncated: boolean | null = null;
  if (truncRaw === '1') truncated = true;
  else if (truncRaw === '0') truncated = false;

  return {
    limit: parseIntHeader(headers, HDR_LIMIT),
    returned: parseIntHeader(headers, HDR_RETURNED),
    truncated,
    nextCursor: headers.get(HDR_NEXT_CURSOR),
    total: parseIntHeader(headers, HDR_TOTAL),
  };
}

/** 用户可见摘要：明确区分「本页」与「总量」。 */
export function formatListCoverage(meta: ListMeta, shown: number): string {
  const page = meta.returned ?? shown;
  if (meta.total != null) {
    if (meta.truncated) {
      return `已显示 ${shown} 条（本页 ${page}），共 ${meta.total} 条；还有更多`;
    }
    return `已显示 ${shown} 条，共 ${meta.total} 条`;
  }
  if (meta.truncated === true) {
    return `已显示 ${shown} 条（本页 ${page}）；列表已截断，条数不是全量`;
  }
  if (meta.truncated === false) {
    return `已显示 ${shown} 条（本页完整，未截断）`;
  }
  return `已显示 ${shown} 条（未提供分页元数据，不得视为全量）`;
}

File: apps/control-api/app/list_meta.py
"""列表分页元数据（DEV13-D）。

合同：
- JSON 体仍可为数组（兼容既有客户端）；
- 用响应头声明本页条数、是否截断、下一游标；
- **禁止**用「本页 length」冒充全量；截断时必须标 truncated。

头字段：
- X-SIQ-List-Limit
- X-SIQ-List-Returned
- X-SIQ-List-Truncated  (0|1)
- X-SIQ-Next-Cursor     (截断时可选)
- X-SIQ-List-Total      (可选；仅在调用方显式计算时设置)
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from starlette.responses import Response

HDR_LIMIT = "X-SIQ-List-Limit"
HDR_RETURNED = "X-SIQ-List-Returned"
HDR_TRUNCATED = "X-SIQ-List-Truncated"
HDR_NEXT_CURSOR = "X-SIQ-Next-Cursor"
HDR_TOTAL = "X-SIQ-List-Total"

# 暴露给浏览器前端（与 CORS 中间件配合时需允许）
EXPOSE_HEADERS = (
    HDR_LIMIT,
    HDR_RETURNED,
    HDR_TRUNCATED,
    HDR_NEXT_CURSOR,
    HDR_TOTAL,
)


def clamp_limit(limit: int, *, default: int = 50, hard_max: int = 200) -> int:
    if limit < 1:
        return default
    return min(limit, hard_max)


def apply_list_meta(
    response: Response,
    *,
    limit: int,
    returned: int,
    truncated: bool,
    next_cursor: str | None = None,
    total: int | None = None,
) -> None:
    response.headers[HDR_LIMIT] = str(limit)
    response.headers[HDR_RETURNED] = str(returned)
    response.headers[HDR_TRUNCATED] = "1" if truncated else "0"
    if truncated and next_cursor:
        response.headers[HDR_NEXT_CURSOR] = next_cursor
    if total is not None:
        response.headers[HDR_TOTAL] = str(total)


def take_page(
    rows_plus_one: Sequence[Any],
    *,
    limit: int,
) -> tuple[list[Any], bool]:
    """对 limit+1 查询结果切页；truncated=True 表示还有后续。"""
    items = list(rows_plus_one)
    if len(items) > limit:
        return items[:limit], True
    return items, False
